qcengine/tests/test_config.py

```diff
@@ -155,8 +164,18 @@ def test_config_local_njob_ncore_plus_memory(opt_state_basic):
 
 
 def test_config_local_nnodes(opt_state_basic):
-    config = qcng.config.get_config(hostname="something", local_options={"nnodes": 4})
-    assert config.nnodes == 4
+    # Give a warning that mentions that mpirun is needed if you define a multi-node task
+    with pytest.raises(ValueError) as exc:
+        qcng.config.get_config(hostname="something", local_options={"nnodes": 10})
+    assert 'https://qcengine.readthedocs.io/en/stable/environment.html' in str(exc.value)
+
+    # Test with an MPI run command
+    config = qcng.config.get_config(hostname="something",
+                                    local_options={"nnodes": 4,
+                                                   "mpirun_command": "mpirun -n {total_ranks} -N {ranks_per_node}"})
+    assert config.use_mpirun
+    assert config.mpirun_command.startswith('mpirun')
+    assert create_mpi_invocation('hello_world', config) == ['mpirun', '-n', '20', '-N', '5', 'hello_world']
```

Comment:
This is a really nice way of handling this!

@taylor-a-barnes What do you think of this kind of automated api config?

Reply:
@dgasmith @WardLT 

Generally speaking, this should work fine.  One thing you should probably plan for is how this api will handle issues related to MPI task pinning, especially in the context of hybrid OMP-MPI calculations.  Some nice examples of these complications are described [at NERSC](https://docs.nersc.gov/jobs/affinity/).  A few specific points to consider: 

1. How will you handle task affinity with respect to logical cores (see, for example, the srun "-cpus-per-task" and "-cpu-bind" options)?  Is it worthwhile to have a logical_cores_per_rank option?
2. Similarly, is it worthwhile to have a ranks_per_socket option for handling socket affinity?
3. How does the user specify a job that involves, for example, 16 MPI ranks with 4 OMP threads per rank? Can the api derive the relevant information from ncores and total_ranks?

I'll also point out that it is generally a best practice to use "mpiexec" rather than "mpirun"; I suggest that you make this change wherever possible.

Reply:
All good questions, @taylor-a-barnes !

My current approach to affinity related issues and handling hybrid parallelism is to allow the users to hard-code them in the `mpirun_command` (e.g., adding a `-cpu-per-task` flag, or a `-depth` ). As I don't have a good grasp of MPI launch systems, I kept the configuration simple.

That said, it would be wonderful to be able to set the number of logical cores per node. What do you think about a `ranks_per_node` configuration variable for `TaskConfig`? That could then be used to compute a `cores_per_rank` variable we can add to the mpirun command.

Also, I'll go ahead and rename things to mpiexec from mpirun

Reply:
@taylor-a-barnes , I've pushed a commit that adds the ability to set the number of "cores/threads per rank"

Reply:
@WardLT Great! This tends to be one of the more overlooked and thus error-prone elements of launching MPI jobs, so I think users will find this helps them avoid a fair number of mistakes.